developing/languages/GO.md

```diff
@@ -4,9 +4,184 @@
 
 Go aka Golang is a programming language designed by folks at Google.
 
-## Resources
+This guide provides some general resources for working in Go.
+[Web Application in Go](../../web/server/go.md)
+provides some specifics to web development.
+
+## Learning Resources
+
+### References
 
 * [The Go Programming Language Official Website](https://golang.org/)
+* [Effective Go](https://golang.org/doc/effective_go.html) (how to do things “the Go way”)
 * [Go Pointer Primer](https://github.com/trussworks/go-pointer-primer)
+* [GoDoc](https://godoc.org/) (where you can read the docs for nearly any Go package)
+* [Go wiki](https://github.com/golang/go/wiki/Learn)
+* _Book_: [The Go Programming Language](http://www.gopl.io/)
+* Advanced Testing with Go
+  [Video](https://www.youtube.com/watch?v=yszygk1cpEc)
+  and [Article](https://about.sourcegraph.com/go/advanced-testing-in-go) (great overview of useful techniques, useful for all Go programmers)
+
+### Tours/Lessons
+
+* [A Tour of Go](https://tour.golang.org) (in-browser interactive language tutorial)
+* [How to Write Go Code](https://golang.org/doc/code.html) (info about the Go environment, testing, etc.)
 * [Go by Example](https://gobyexample.com)
 * [Go Track on Exercism](https://exercism.io/tracks/go)
+
+## Testing
+
+### General
+
+* Use table-driven tests where appropriate.
+* Make judicious use of helper functions so that the intent of a test is not lost in a sea of error checking and boilerplate.
+  Use [`t.Helper()`](https://golang.org/pkg/testing/#T.Helper) in your test helper functions to keep stack traces clean.
+
+### Coverage
+
+* Always test exported functions.
+  Exported functions should be treated as an API layer for other packages.
+  Cover the expected behavior and error scenarios as a user of that API.
+* Try not to test unexported functions.
+  Unexported functions are implementation details of exported ones
+  and should not change the intended usage.
+  If you find that an unexported function is complex and needs testing,
+  it might mean it needs to be refactored as it's exported function elsewhere.
```

Comment:
If we want to enforce black box testing, should we have a rule that tests should be in a separate `*_test` package and not in the same package as the code being tested?  ([reference](https://stackoverflow.com/questions/19998250/proper-package-naming-for-testing-with-the-go-language))

Reply:
I think it's _ok_ in some scenarios, but general pattern should be to think about why that's happening. I think long, sequential algorithms often amount to private method testing, which is reasonable. But generally, if code is getting super deep, passing in dependencies that are tested separate seems cleaner to me.

Integration tests should fall into a separate package though. This is not a distinction we've made in the past. Something that can probably be addressed later in this doc.